add Isotonic Regression

presnick opened the pull request:
Also, a notebook, Alterante_Calibration_Curves, that puts all the alternative calibration curves through their paces.

presnick committed:
Add isotonic regression calibration curve

presnick committed:
Merge remote-tracking branch 'siqi/main'

avalanchesiqi commented on pyquantifier/data.py:
```diff
@@ -327,6 +328,22 @@ def generate_calibration_curve(self, method='platt scaling', num_bin=10):
+
+            # Fit isotonic regression model
+            ir = IsotonicRegression(out_of_bounds='clip')
+            y_axis = ir.fit_transform(sorted_train_CX, sorted_train_GT)
```
I think this should not be `.fit_transform()`

Based on the [doc](https://scikit-learn.org/dev/modules/generated/sklearn.isotonic.IsotonicRegression.html#sklearn.isotonic.IsotonicRegression.fit_transform) of Isotonic Regression, 
> Fits transformer to X and y with optional parameters fit_params and returns a transformed version of X.

I think we should always use the same x-axis (i.e., `np.arange(0.5/100, 1, 1/100)`) for Platt Scaling and Isotonic Regression. `sorted_train_CX` has `num_item` of the base dataset in it, usually at the level of tens of thousands.

I have fixed it in the latest commit.

presnick replied:
I think we would only run .fit_transform() on the items in the training set, so not tens of thousands. But still, I get your point that we don't need all of the tiny jump points to make a good plot.

Actually, I was just thinking that the IsotonicRegresion instance, after we call `.fit()`, has a built-in `.predict()` method.

Why not store that instance inside our IsotonicRegressionCalibrationCurve instance, and use it to implement the `.get_calibrated_prob()` method that's needed in that subclass? Likely to be more efficient and less bug-prone than  what I implemented.

For plotting, you could do as you are suggesting: generate the 100 x-axis values like you do for Platt Scaling and pass them through `.get_calibrated_prob()` to get corresponding y-axis values.

avalanchesiqi replied:
> I was just thinking that the IsotonicRegresion instance, after we call .fit(), has a built-in .predict() method.

This is indeed what I implemented yesterday. I didn't store the instance. Instead, I store the `x_axis` and corresponding `y_axis` for the 100 bins. The reason is that for any of `nonparametric binning, piecewise linear, adjusted piecewise linear`, they don't have a built-in instance. They are defined by a `x_axis` and a `y_axis`. I want to keep it consistent.

Of course I could change the stored object for `platt scaling, isotonic regression` to the fitted instance. The `.get_calibrated_prob()` method will become a wrapper of `.predict()` or `.predict_proba()`.

What do you think? Having a universal `x_axis` and `y_axis` design for all calibration curves or `fitted_instance` for `platt scaling, isotonic regression` and `x_axis` and `y_axis` for `nonparametric binning, piecewise linear, adjusted piecewise linear`?

presnick commented:
I like the idea of wrapping the built-in `.predict()` or `.predict_proba()`
for platt and isotonic, rather than reimplementing anything, and certainly
better than a reimplementation that loses a little accuracy by binning.

We could still have a consistent way of doing plots, with the x_axis and
y_axis values stored. But even there, I ended up implementing .plot() in
the subclasses for binned curves, because sometimes we want linear
interpolations and sometimes we just want step functions.

On Wed, Nov 20, 2024 at 11:55 AM Siqi Wu ***@***.***> wrote:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In pyquantifier/data.py
> <https://github.com/avalanchesiqi/pyquantifier/pull/3#discussion_r1850679359>
> :
>
> > @@ -327,6 +328,22 @@ def generate_calibration_curve(self, method='platt scaling', num_bin=10):
>              return prob_cali_obj
>          elif method == 'temperature scaling':
>              pass
> +        elif method == 'isotonic regression':
> +            train_CX = self.df['p_pos'].values
> +            train_GT = self.df['gt_label'].map({'neg': 0, 'pos': 1}).values
> +
> +            # Sort the train_CX and train_GT arrays
> +            sorted_indices = np.argsort(train_CX)
> +            sorted_train_CX = train_CX[sorted_indices]
> +            sorted_train_GT = train_GT[sorted_indices]
> +
> +            # Fit isotonic regression model
> +            ir = IsotonicRegression(out_of_bounds='clip')
> +            y_axis = ir.fit_transform(sorted_train_CX, sorted_train_GT)
>
> I was just thinking that the IsotonicRegresion instance, after we call
> .fit(), has a built-in .predict() method.
>
> This is indeed what I implemented yesterday. I didn't store the instance.
> Instead, I store the x_axis and corresponding y_axis for the 100 bins.
> The reason is that for any of nonparametric binning, piecewise linear,
> adjusted piecewise linear, they don't have a built-in instance. They are
> defined by a x_axis and a y_axis. I want to keep it consistent.
>
> Of course I could change the stored object for platt scaling, isotonic
> regression to the fitted instance. The .get_calibrated_prob() method will
> become a wrapper of .predict() or .predict_proba().
>
> What do you think? Having a universal x_axis and y_axis design for all
> calibration curves or fitted_instance for platt scaling, isotonic
> regression and x_axis and y_axis for nonparametric binning, piecewise
> linear, adjusted piecewise linear?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/avalanchesiqi/pyquantifier/pull/3#discussion_r1850679359>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAIG7TQWPUV2YKDBWCAALK32BS5ITAVCNFSM6AAAAABSAUZAW6VHI2DSMVQWIX3LMV43YUDVNRWFEZLROVSXG5CSMV3GSZLXHMZDINBZGIYTCMRYGM>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>

avalanchesiqi commented:
Wait, for the binned curves, shouldn't the linear interpolation and step function also have an effect on the `. get_calibrated_prob()` method? For instance, the step function will have the same `y_value` for `x \in [0, 0.1]`, but the linear interpolation will have different `y_value`.

I think the `.plot()` should be the same for all CalibrationCurve, because it just plots the `x_axis` and `y_axis` attributes.

presnick commented:
Yes, I agree that the linear interpolation methods should have a different way of calculating `.get_calibrated_prob()`.

The generic `.plot()` you had before draws lines connecting adjacent (x-axis, y-axis) points. So it won't quite display as a step-function, even if you get 100 x-axis points. It may be close enough, being flat in the rang [0, 0.1]. But there will be a linear interpolation between [0.1, 0.11]. Maybe that wouldn't be visible to the naked eye, so isn't worth worrying about.

The other issue is that we are kind of overload what `.x-axis` means with binned calibration curved. Are they the points to plot using a generic .plot() function? Or are they the points used to compute `.get_calibrated_prob()` in those cases where we're not wrapping a scikit learn object.

Maybe the right solution is that when we don't have a scikit learn object that we're wrapping, we pass in something other than `.x-axis` and `.y-axis` to compute calibrated probabilities from. And then we always fill `.x-axis` and `.y-axis` with 100 pairs, with y-axis values computed by calling  `.get_calibrated_prob()`
